hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/index/IndexResolver.java

```diff
@@ -691,41 +822,61 @@ private static MapIndexScanPhysicalRel createIndexScan(
             filter,
             converterTypes,
             exp,
-            remainderExp
+            remainderExp,
+            descending
         );
     }
 
     /**
-     * Create an index scan without any filter. Used by HD maps only.
+     * Builds a collation with collation fields re-mapped according with the table projects
      *
-     * @param scan the original scan operator
+     * @param scan       the logical map scan
+     * @param index      the index
+     * @param descending whether the collation is descending
+     * @return the new collation trait
+     */
+    private static RelCollation buildCollationTrait(MapScanLogicalRel scan,
+                                                    MapTableIndex index,
+                                                    boolean descending) {
+        List<RelFieldCollation> fields = new ArrayList<>(index.getFieldOrdinals().size());
+        HazelcastTable table = scan.getTableUnwrapped();
+        for (Integer indexFieldOrdinal : index.getFieldOrdinals()) {
+            int remappedIndexFieldOrdinal = table.getProjects().indexOf(indexFieldOrdinal);
+            if (remappedIndexFieldOrdinal == -1) {
+                // The field is not used in the query
+                break;
+            }
+            RelFieldCollation.Direction direction = descending ? RelFieldCollation.Direction.DESCENDING
+                : RelFieldCollation.Direction.ASCENDING;
+            RelFieldCollation fieldCollation = new RelFieldCollation(remappedIndexFieldOrdinal, direction);
+            fields.add(fieldCollation);
+        }
+        return RelCollations.of(fields);
+    }
+
+    /**
+     * Creates an index scan without any filter.
+     *
+     * @param scan         the original scan operator
      * @param distribution the original distribution
-     * @param indexes available indexes
+     * @param index        available indexes
      * @return index scan or {@code null}
      */
     public static RelNode createFullIndexScan(
```

Comment:
Looks like this method could be private.

Reply:
Fixed in https://github.com/hazelcast/hazelcast/pull/18013/commits/c2468be2ca3db367a1dd1cc79b219b0dcf807370